Charts

HugoGoHe opened the pull request:
Worked on the Stats page. 

- Body weight change over time irregular intervals chart.
- Weight progression in certain exercise basic line chart.
- Prediction using logarithmic regression.

HugoGoHe committed:
[Slideshow page][slideshow]

- New button on the WeightLoss page to navigate to the Slideshow page.

- An array is passed with the urls of the images to load them in a slideshow.

- By implementing the KASlideShow library the user can see a slideshow of their photos, they can pause, resume, tap the image to go to the next one and adjust the speed of the transitions.

HugoGoHe committed:
[Slideshow page][WeightLoss page][Cleaned the code]

HugoGoHe committed:
[Post Preview page][Input validation]

- If user does not enter a valid weight an error label is displayed and the keyboard is dismissed.

HugoGoHe committed:
[CurrentWorkout page][Bug fixed]

- Now the tableView is displayed only if suggestedExercises.count > 0.

HugoGoHe committed:
[Stats_page][Body weight change over time chart]

-Created an irregular intervals chart that shows the user's body weight change over time.


Extra:
-Changed PFObject name: Post to ProgressPic
-Changed range of delay between transitions on the slideshow

HugoGoHe committed:
[Stats page][Weight lifted in exercises chart]

- Added the following properties to PFObject Exercise:
    -username
    -postedAt

- TableView loads charts of the weight progression of the exercises the user has performed.

HugoGoHe committed:
[Stats page][Reloading data]

- The tableview's data is loaded correctly
- Now the user is firstly directed to the Workouts page

HugoGoHe committed:
[Stats page][Logarithmic trendline]

- A logarithmic trendline is being graphed for every "Weight lifted in exercise" chart. This was made by using Logarithmic regression.

HugoGoHe committed:
[Stats page][Bug: reloading charts]

Bug solved:

The app crashed when the user pulled to refresh the charts .  The mutableArrays never had their objects removed. This is solved using the method removeAllObjects inside the completion block of the second query.

HugoGoHe committed:
[Stats page][Optimized logarithmicRegression function]

Using only counters "i" instead of making the X values array.

HugoGoHe committed:
[Post preview page][Resizing images]

- Added function to resize the images used by the user in order for them to be an acceptable filie size for the Parse database.

HugoGoHe committed:
[Cleaned code][Commented code]

cmnagarajan commented on Capstone Project/SlideshowViewController.h:
```diff
@@ -0,0 +1,19 @@
+#import "KASlideShow.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
```
declare 
@protocol KASlideShowDelegate;
and remove #import "KASlideShow.h"

cmnagarajan commented on Capstone Project/SlideshowViewController.m:
```diff
@@ -0,0 +1,88 @@
+    [_speedSlider setUserInteractionEnabled:NO];
+
+     _datasource = self.urls;
+    NSLog(@"%@", self.uiImages);
```
Remove this log statement before commit

cmnagarajan commented on Capstone Project/ViewControllers/WeightLoss/WeightLossViewController.m:
```diff
@@ -121,5 +127,13 @@ - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+        UINavigationController *nav = [segue destinationViewController];
+        SlideshowViewController *ssvc = (SlideshowViewController *) nav.topViewController;
+        ssvc.urls = self.urls;
+        NSLog(@"%@", ssvc.urls);
```
Remove NSLog

cmnagarajan commented on Capstone Project/SlideshowViewController.m:
```diff
@@ -0,0 +1,88 @@
+    
+    [_slideshow setTransitionType:KASlideShowTransitionFade];
+    _slideshow.gestureRecognizers = nil;
+    [_slideshow addGesture:KASlideShowGestureTap];
```
It is repeating the code in line 42

cmnagarajan replied:
Why switching between self.slideShow and _slideshow. Please make it consistent

cmnagarajan commented on Capstone Project/SlideshowViewController.m:
```diff
@@ -0,0 +1,88 @@
+
+- (NSObject *)slideShow:(KASlideShow *)slideshow objectAtIndex:(NSUInteger)index
+{
+    return _datasource[index];
```
To avoid crash please do this
if (index < _datasource.count) {
 return _datasource[index];
}

HugoGoHe replied:
The function always need to return something, even if it doesn't go through the if statement, I already tested it without any photos and it doesn't crash

cmnagarajan commented on Capstone Project/ViewControllers/WeightLoss/PostPreviewViewController.m:
```diff
@@ -27,17 +28,26 @@ - (void)viewDidLoad {
-    [nav setSelectedViewController:[nav.viewControllers objectAtIndex:0]];
-    [self.navigationController presentViewController:nav animated:YES completion:nil];
+    if (!([self.weightLabel.text intValue] > 0 && [self.weightLabel.text intValue] < 1000)){
+        self.errorLabel.text = @"Type a valid weight";
```
Use placeholder method
self.errorLabel.placeholder = @"Type a valid weight";

HugoGoHe replied:
<img width="832" alt="image" src="https://user-images.githubusercontent.com/88735772/183148211-4878d680-7031-43c1-ad85-e24a1d400d02.png">
isn't placeholder for UITextFields?

cmnagarajan commented on Capstone Project/ViewControllers/Stats/StatsViewController.m:
```diff
@@ -26,14 +42,69 @@ - (void)viewDidLoad {
 
-    [self createChart:self.WeightChartView];
+        }else{
+            NSLog(@"Error: %@ %@", error, [error userInfo]);
```
Remove NSLog statements

cmnagarajan commented on Capstone Project/ViewControllers/Stats/StatsViewController.m:
```diff
@@ -26,14 +42,69 @@ - (void)viewDidLoad {
+            for (int i = 0; i < self.dates.count; i++) {
+                [self.formatedDates addObject:[dateFormatter stringFromDate:self.dates[i]]];
+            }
+       //     NSLog(@"%@", self.formatedDates);
```
Remove commented code

cmnagarajan commented on Capstone Project/ViewControllers/Stats/StatsViewController.m:
```diff
@@ -42,7 +113,169 @@ - (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(N
+    
+    NSMutableArray *arrayOfPoints = [[NSMutableArray alloc] init];
+    for (int i=0; i<self.weights.count; i++) {
+        //        NSMutableArray *point = [[NSMutableArray alloc] init];
```
remove commented code

cmnagarajan commented on Capstone Project/ViewControllers/Stats/StatsViewController.m:
```diff
@@ -42,7 +113,169 @@ - (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(N
+        
+        NSNumber *unixtimeMilisecondsObject = [NSNumber numberWithInteger:unixtimeMiliseconds];
+        
+        NSMutableArray * point = [[NSMutableArray alloc] init];
```
You can avoid creating a mutable array when the objects to add is already known
NSArray * point = [NSArray arrayWithObjects:unixtimeMilisecondsObject, [self.weights objectAtIndex:i], nil];

cmnagarajan commented on Capstone Project/ViewControllers/Stats/StatsViewController.m:
```diff
@@ -42,7 +113,169 @@ - (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(N
+    spline1.name = @"Time";
+    spline1.data = arrayOfPoints;
+    
+    //     spline1.data = @[
```
Remove this

cmnagarajan replied:
Commiting before cleaning up the code is not the right thing to do

cmnagarajan commented on Capstone Project/ViewControllers/Stats/StatsViewController.m:
```diff
@@ -24,8 +24,8 @@ @interface StatsViewController () <UITableViewDelegate, UITableViewDataSource>
 @property (strong, nonatomic) NSArray *availableExercises;
-
-
+@property (strong, nonatomic) NSArray *userExercises;
```
Declare the object type

cmnagarajan commented on Capstone Project/ViewControllers/Stats/StatsViewController.m:
```diff
@@ -52,6 +52,10 @@ - (void)viewDidLoad {
     self.weights = [[NSMutableArray alloc] init];
     self.formatedDates = [[NSMutableArray alloc] init];
+    
+    self.availableExercises = [[NSArray alloc] init];
```
It seems the intent here is to use it as a NSArray pointer only. No need to allocate

cmnagarajan commented on Capstone Project/ViewControllers/Stats/StatsViewController.m:
```diff
@@ -301,5 +306,54 @@ - (void)basicLineChart:(UIView *) cellView ofExercise:(NSString *)nameOfExercise
 }
 
 
+-(NSMutableArray *)logarithmicRegression:(NSMutableArray *)Y{
```
Please add some detailed comments on what is implemented here step by step

cmnagarajan commented on Capstone Project/ViewControllers/Stats/StatsViewController.m:
```diff
@@ -51,16 +51,18 @@ - (void)viewDidLoad {
 }
 
--(void)getData{
+-(void)getAvailableExercises{
```
-(void)availableExercises{
Using get prefix on getter methods is not Objc convention

cmnagarajan commented on Capstone Project/ViewControllers/Stats/StatsViewController.m:
```diff
@@ -83,6 +85,12 @@ -(void)getData{
+
+
+
+-(void)getData{
```
Same as above

SpencerLerner97 commented on Capstone Project/ViewControllers/Stats/StatsViewController.m:
```diff
@@ -351,10 +347,8 @@ -(NSMutableArray *)logarithmicRegression:(NSMutableArray *)Y{
-        
-//        prediction = A + B * log([X[i] doubleValue]);
-        prediction = A + B * log(i + 1);
+    for(int i = 1; i < ((n+1) + 3); i++){
```
Very cool, so if I understand log regression its just y = m*log(x) + b as opposed to y = mx + b?

HugoGoHe replied:
that's right

HugoGoHe committed:
[PR Feedback]

- Declared:
@protocol KASlideShowDelegate;
and removed: 
#import "KASlideShow.h"

- Refactored slideshow page code

- Now self.availableExercises is not being allocated

- Changed names of functions that started with "get"

- Replaced NSLogs that printed error descriptions with alerts.

- Now point is an NSArray instead of an NSMutableArray
